flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierUnaligner.java

```diff
@@ -0,0 +1,417 @@
+		 * received, all new buffers need to be persisted after unaligned checkpointing has been started.
+		 */
+		private final boolean[] barrierReceivedChannels;
+
+		/**
+		 * The number of input channels which has read the barrier by task.
+		 */
+		private int numBarriersReceived;
+
+		/** A future indicating that all barriers of the a given checkpoint have been read. */
+		private CompletableFuture<Void> allBarriersReceivedFuture = FutureUtils.completedVoidFuture();
+
+		/**
+		 * The checkpoint id to guarantee that we would trigger only one checkpoint when reading the same barrier from
+		 * different channels.
+		 *
+		 * <p>Note: this checkpoint is valid in respect to <b>received</b> barriers in contrast to
+		 * {@link CheckpointBarrierUnaligner#currentCheckpointId}.
+		 */
+		private long currentCheckpointId = -1L;
+
+		/** The number of opened channels. */
+		private int numOpenChannels;
+
+		private final ChannelStateWriter channelStateWriter;
+
+		private final CheckpointBarrierUnaligner handler;
+
+		public ThreadSafeUnaligner(
+				int totalNumChannels,
+				ChannelStateWriter channelStateWriter,
+				CheckpointBarrierUnaligner handler) {
+			barrierReceivedChannels = new boolean[totalNumChannels];
+			this.channelStateWriter = channelStateWriter;
+			this.handler = handler;
+			Arrays.fill(barrierReceivedChannels, true);
+			numOpenChannels = totalNumChannels;
+		}
+
+		@Override
+		public synchronized void notifyBarrierReceived(CheckpointBarrier barrier, InputChannelInfo channelInfo) throws IOException {
+			long barrierId = barrier.getId();
+
+			if (currentCheckpointId < barrierId) {
+				handleNewCheckpoint(barrierId);
+				handler.runInTaskThread(() -> handler.notifyCheckpoint(barrier, 0, 0), "notifyCheckpoint");
+			}
+
+			int channelIndex = handler.getFlattenedChannelIndex(channelInfo);
+			if (barrierId == currentCheckpointId && !barrierReceivedChannels[channelIndex]) {
+				if (LOG.isDebugEnabled()) {
+					LOG.debug("{}: Received barrier from channel {} @ {}.", handler.taskName, channelIndex, barrierId);
+				}
+
+				barrierReceivedChannels[channelIndex] = true;
+
+				if (++numBarriersReceived == numOpenChannels) {
+					allBarriersReceivedFuture.complete(null);
+				}
+			}
+		}
+
+		@Override
+		public synchronized void notifyBufferReceived(Buffer buffer, InputChannelInfo channelInfo) {
+			if (!barrierReceivedChannels[handler.getFlattenedChannelIndex(channelInfo)]) {
+				channelStateWriter.addInputData(
+					currentCheckpointId,
+					channelInfo,
+					ChannelStateWriter.SEQUENCE_NUMBER_UNKNOWN,
+					buffer.retainBuffer());
+			}
+		}
+
+		@Override
+		public void close() throws IOException {
+			allBarriersReceivedFuture.completeExceptionally(new IllegalStateException("Closed CheckpointBarrierUnaligner"));
+		}
+
+		private synchronized void handleNewCheckpoint(long barrierId) throws IOException {
+			if (!allBarriersReceivedFuture.isDone()) {
```

Comment:
Good idea. I just fear that it may lead to the checkpoint being aborted with the provided exception instead of the `CHECKPOINT_DECLINED_SUBSUMED` failure message from `notifyAbort`. Could you please double-check?